public/app/features/plugins/extensions/utils.tsx

```diff
@@ -421,3 +425,221 @@ export function getLinkExtensionPathWithTracking(pluginId: string, path: string,
+    .filter(
+      (app) =>
+        app.extensions.addedLinks.some((link) => link.targets.includes(extensionPointId)) ||
+        app.extensions.addedComponents.some((component) => component.targets.includes(extensionPointId))
+    )
+    .map((app) => app.id)
+    .reduce((acc: string[], id: string) => {
+      return [...acc, id, ...getAppPluginDependencies(id)];
+    }, []);
+};
+
+// Returns a list of app plugin ids that are necessary to be loaded to use the exposed component.
+// (It is first the plugin that exposes the component, and then the ones that it depends on.)
+export const getExposedComponentPluginDependencies = (exposedComponentId: string) => {
+  const pluginId = getAppPluginIdFromExposedComponentId(exposedComponentId);
+
+  return [pluginId].reduce((acc: string[], pluginId: string) => {
+    return [...acc, pluginId, ...getAppPluginDependencies(pluginId)];
+  }, []);
+};
+
+// Returns a list of app plugin ids that are necessary to be loaded, based on the `dependencies.extensions`
+// metadata field. (For example the plugins that expose components that the app depends on.)
+// Heads up! This is a recursive function.
+export const getAppPluginDependencies = (pluginId: string): string[] => {
+  if (!config.apps[pluginId]) {
+    return [];
+  }
+
+  const pluginIdDependencies = config.apps[pluginId].dependencies.extensions.exposedComponents.map(
+    getAppPluginIdFromExposedComponentId
+  );
+
+  return pluginIdDependencies.reduce((acc, pluginId) => {
+    return [...acc, ...getAppPluginDependencies(pluginId)];
+  }, pluginIdDependencies);
+};
+
+// Returns a list of app plugins that has to be loaded before core Grafana could finish the initialization.
+export const getAppPluginsToAwait = () => {
+  const pluginIds = [
+    // The "cloud-home-app" is registering banners once it's loaded, and this can cause a rerender in the AppChrome if it's loaded after the Grafana app init.
+    'cloud-home-app',
+  ];
+
+  return Object.values(config.apps).filter((app) => pluginIds.includes(app.id));
+};
+
+// Returns a list of app plugins that has to be preloaded in parallel with the core Grafana initialization.
+export const getAppPluginsToPreload = () => {
+  // The DashboardPanelMenu extension point is using the `getPluginExtensions()` API in scenes at the moment, which means that it cannot yet benefit from dynamic plugin loading.
+  const dashboardPanelMenuPluginIds = getExtensionPointPluginDependencies(PluginExtensionPoints.DashboardPanelMenu);
+  const isLazyLoadingEnabled = config.featureToggles.appPluginLazyLoading;
+  const awaitedPluginIds = getAppPluginsToAwait().map((app) => app.id);
+
+  if (!isLazyLoadingEnabled) {
+    return Object.values(config.apps)
+      .filter((app) => app.preload)
+      .filter((app) => !awaitedPluginIds.includes(app.id));
+  }
+
+  return Object.values(config.apps)
+    .filter((app) => app.preload)
+    .filter((app) => !awaitedPluginIds.includes(app.id))
+    .filter(
```

Comment:
Usually don't care about these things, but since this happens at app bootstrap you may want to replace these three loops with a simple for loop to improve performance.

Reply:
That's fair enough. I have updated it in https://github.com/grafana/grafana/pull/86624/commits/c131a5d59ae6c02e98ff51bbdb23509124c1d784 (instead of a for loop I kept using the `.filter()` method so I can "just return" with the results, but in this sense I think they are identical, they both loop through the list once.